pkg/plugin/sdk/source.go

```diff
@@ -0,0 +1,261 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sdk
+
+import (
+	"context"
+	"io"
+	"sync/atomic"
+	"time"
+
+	"github.com/conduitio/conduit-plugin/cpluginv1"
+	"github.com/conduitio/conduit/pkg/foundation/cerrors"
+	"github.com/jpillora/backoff"
+	"gopkg.in/tomb.v2"
+)
+
+// Source TODO
+// All implementations must embed UnimplementedSource for forward compatibility.
+type Source interface {
+	// Configure is the first function to be called in a plugin. It provides the
+	// plugin with the configuration that needs to be validated and stored. In
+	// case the configuration is not valid it should return an error.
+	Configure(context.Context, map[string]string) error
+
+	// Open is called after Configure to signal the plugin it can prepare to
+	// start producing records. If needed, the plugin should open connections in
+	// this function.
+	Open(context.Context, Position) error
+
+	// Read returns a new Record and is supposed to block until there is either
+	// a new record or the context gets cancelled. It can also return the error
+	// ErrBackoffRetry to signal to the SDK it should call Read again with a
+	// backoff retry.
+	// It's guaranteed that Read will be called with a cancelled context at
+	// least once, or that the context will get cancelled while Read is running.
+	// In that case Read must stop retrieving new records from the source system
+	// and start returning records that may have already been buffered. If there
+	// are no buffered records left Read must return the context error to signal
+	// a graceful stop.
+	// After Read returns an error the function won't be called again (except if
+	// the error is ErrBackoffRetry, as mentioned above).
+	// Read can be called concurrently with Ack.
+	Read(context.Context) (Record, error)
```

Comment:
The consequence is that Conduit can't access a record at a random position, but we don't have a use case for that anyway. It simplifies the life of the plugin developer because they don't have to think about random access. So yeah, I also think this is a subtle but important change.

Reply:
Not having to access messages at random offsets in Kafka made it way simpler in the Kafka plugin, so much I can say.: )